File: 1123.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def lcaDeepestLeaves(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        def postorder(node) -> tuple[Optional[TreeNode], int]:
            if node is None:
                return (None, -1)
            (leftRes, maxLeftDep) = postorder(node.left)
            (rightRes, maxRightDep) = postorder(node.right)
            if maxLeftDep == maxRightDep:
                return (node, maxLeftDep + 1)
            else:
                return (leftRes, maxLeftDep + 1) \
                    if maxLeftDep > maxRightDep \
                    else (rightRes, maxRightDep + 1)

        return postorder(root)[0]

File: sol/1123.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
struct Solution {
    TreeNode* postorder(TreeNode *node, int &dep){
        if (node == nullptr){
            dep = -1;
            return nullptr;
        }
        int leftDep, rightDep;
        TreeNode *leftRes = this->postorder(node->left, leftDep), 
                 *rightRes = this->postorder(node->right, rightDep);
        if (leftDep == rightDep){
            dep = leftDep + 1;
            return node;
        } else if (leftDep > rightDep){
            dep = leftDep + 1;
            return leftRes;
        } else {
            dep = rightDep + 1;
            return rightRes;
        }
    }

    TreeNode* lcaDeepestLeaves(TreeNode* root) {
        int dep;
        return this->postorder(root, dep);
    }
};

